Add docs for updating version of GOV.UK Frontend (#1463)
alphagov/govuk-frontend: docs/installation/updating-with-npm.md

# Updating GOV.UK Frontend with node package manager (NPM)

Follow these instructions if you have previously installed the GOV.UK Frontend npm package following the [installation instructions](https://github.com/alphagov/govuk-frontend/blob/master/docs/installation/installing-with-npm.md#requirements), and want to update to a more recent version.

## Steps to update

### 1. Find out changes introduced in recent versions.

> To do this, view `package.json` at the root of your project to find out the version of govuk-frontend that your app is currently running, for instance "2.11.0". See [example](https://github.com/alphagov/govuk-design-system/blob/master/package.json#L34) for where this information lives in GOV.UK Design System.

Reviewer: If we want to include an example could consider inlining a snippet rather than pointing to a file that can change.

However, If we keep pointed to the design system code base it should be permalinked.

Reviewer 2: What about asking people to run `npm list govuk-frontend` instead?

```
$ npm list govuk-frontend
govuk_design_system@ /Users/oliverbyford/Code/govuk-design-system
└── govuk-frontend@2.11.0
```

Author: Thanks @36degrees. We decided to keep both methods in case someone doesn't have access to the terminal.

Review [Release notes](https://github.com/alphagov/govuk-frontend/releases) to see what changes were introduced since this version was released.

### 2. Choose the version to update to.

You can update to the latest release, or to an earlier version.

If you update to the last non-breaking version of GOV.UK Frontend, you shouldn't need to make any changes to your app. If you update to a breaking release, you might need to make changes to your app to make it work correctly with GOV.UK Frontend. [Release notes](https://github.com/alphagov/govuk-frontend/releases) will have information in the "Breaking changes" section of that release about the changes you might need to make.

### 3. Update the version

Run

```
npm install govuk-frontend@2.13.0
```

where 2.13.0 is the version you require.

## Find out about new releases

> You can watch [govuk-frontend](https://github.com/alphagov/govuk-frontend) for "Releases only" to get notified of new releases.

Reviewer: Could we add a screenshot here? Not sure people will know how to do this.

# Updating GOV.UK Frontend using node package manager (NPM)

> You can update with node package manager (NPM) if you [originally installed

Reviewer 3: What's the reasoning behind the "NPM" abbreviation?
The product is branded as "npm" (https://www.npmjs.com/) so we could probably do away with the all caps abbreviation.
Also since the product is branded as "npm", I'm wondering whether we should have it as "node package manager, npm" or similar 🤔

Reviewer: Since the audience for this guidance is web developers, the majority of them will not know npm by it's extended name.

The tool itself is nearly always referred to as `npm`, for example when you use it `npm install`

You can see on Google Trends when you compare the two, there is nearly no activity for the extended version. https://trends.google.com/trends/explore?geo=GB&q=npm,node%20package%20manager

So in this case I would only expand it once and then always use 'npm' for the rest of the guidance.

It's also worth nothing that npm should be lowercase not uppercase.

Reviewer 4: Thanks both. Sounds like we're in agreement - expand the first time (keeping to style), use lowercase 'npm' (as that's the name). I'll fix that up.

@bravegrape, the brackets/comma question is an interesting one! I think our style is still brackets even if the acronym is the known name / product name. A similar-ish GOV.UK example is https://www.gov.uk/make-a-sorn - most people know it as SORN and definitely not 'Statutory Off Road Notification', but we keep the brackets. What do you think?

Reviewer 3: Sounds good to me - let's follow the established pattern 👍

GOV.UK Frontend with NPM](https://github.com/alphagov/govuk-frontend/blob/master/docs/installation/installing-with-npm.md#requirements).

## Find out which version you're using

> You can find out which version of GOV.UK Frontend your project is using with:

Reviewer 3: ```suggestion
You can find out which version of GOV.UK Frontend your project is using by running:
```
Just a suggestion - it's a common docs pattern for commands that need to be run in the Terminal.

Reviewer 4: Thanks Andrea - amended.

> # Updating GOV.UK Frontend with node package manager (NPM)

Reviewer 2: I don't know why, but '_with_ node package manager' just doesn't sound quite right to me. Have you considered 'using' instead?

```suggestion
# Updating GOV.UK Frontend using node package manager (NPM)
```

Reviewer 4: Good point - I'll tweak this.

You can update with node package manager (NPM) if you [originally installed
GOV.UK Frontend with NPM](https://github.com/alphagov/govuk-frontend/blob/master/docs/installation/installing-with-npm.md#requirements).

## Find out which version you're using

You can find out which version of GOV.UK Frontend your project is using with:

```shell
npm list govuk-frontend
```

If you do not have command line access, you can see the version number in the
`package.json` file in the root of your project directory. For example:

```json
'gov-uk frontend': ‘2.11.0’
````

## Update

You can check the [release
notes](https://github.com/alphagov/frontend/releases) in the GOV.UK Frontend Github repository to find out the latest version of GOV.UK Frontend.

You can update to any version.

> You'll need to progressively update to each version following your current

Reviewer 2: This reads to me as though if e.g. you're on `v2.1` and you want to update to `v2.4`, you have to first update to `v2.2`, then `v2.3` and then finally `v2.4`, which shouldn't be the case.

What is it we're trying to convey here?

> version. You will not need to make code changes for GOV.UK Frontend to keep working in

Reviewer 2: Do we want to talk about semantic versioning here? It _might_ be easier for the user if they know that e.g. a 2.x to 2.y change will not include breaking changes, rather than requiring them to read the release notes to find that out?

Reviewer 4: So something like: "You will not need to make code changes for GOV.UK Frontend to keep working in your project, unless you're updating to a major version - for example 2.0 rather than 2.13.0"?

Reviewer: Could I propose for our style that we avoid 

'major', 'minor', 'patch'

and instead always use

'breaking', 'feature', 'fix'?

I think these are more understandable and reflect the CHANGELOG

Reviewer 2: I like that idea a lot when we're talking about an individual release (e.g. 3.0.0 is a 'breaking release'), but I don't know how well it works when we're talking about how _version numbers_ change?

For example, someone updating from v2.13.0 to v3.1.1 is updating to a 'fix' release, but because the major version number changes they know there'll be breaking changes anyway.

We could talk about 'if the major version number changes' (having explained that term somewhere) but I'm not sure we'd ever say 'if the breaking version number changes'

Reviewer: I think that's just a matter of how it's worded specifically what I was responding to:

> unless you're updating to a major version

would be

> unless you're updating to a version with breaking changes

etc

something to talk about later probs

Reviewer 4: How about combining the two, so we clarify it for anyone who might not be clear on either term? Something like: "...unless you're updating to a major version. A version is major if it has a breaking changes section in its release notes."

Reviewer 2: >"...unless you're updating to a major version. A version is major if it has a breaking changes section in its release notes."

It's slightly more nuanced than that though, because the specific version they're updating to is irrelevant unless they happen to be running the immediate preceding release. In the example I gave above, going from v2.13.0 to v3.1.1 is _not_ 'updating to a major version' (the version they're updating to is a patch release!) but the major version number does change.

I guess we need to work out what's easier for the user – going through the release notes to understand whether there was a breaking change _at some point_, or understanding the basics of how major, minor and patch version numbers work?

Reviewer 4: How about this (to cover this discussion and Ollie's comment about 'progressive' updates too):

You will need to make code changes to make sure GOV.UK keeps working in your project, if you’re updating:

- to a major version - for example you’re updating to 2.0.0
- past a major version - for example you’re updating from 1.2.0 to 2.3.0

A version is major ('breaking') if there's a 'Breaking changes' section in the release notes.

Reviewer 2: That sounds ace, thanks @m-green 👏 

I wonder if it should say 'You *may* need to make code changes' (sometimes, the 'breaking' part of the release might not affect a user) and whether we need to point the user back to the release notes so that they know what those code changes might be?

your project, unless the release note for the version has a ‘Breaking changes’
section.

To update, run:

```shell

> npm install govuk-frontend@2.13.0

Reviewer 2: Should we use `@latest` here, as that'll be the most common use case?

The sentence below could then be something like:

> If you want to install a specific version, replace `@latest` with the version that you want to update to, for example `npm install govuk-frontend@2.13.0`

Reviewer 4: Sounds good - will amend.

# Updating GOV.UK Frontend using node package manager (NPM)

You can update with node package manager (NPM) if you [originally installed
GOV.UK Frontend with NPM](https://github.com/alphagov/govuk-frontend/blob/master/docs/installation/installing-with-npm.md#requirements).

## Find out which version you're using

You can find out which version of GOV.UK Frontend your project is using with:

```shell
npm list govuk-frontend
```

If you do not have command line access, you can see the version number in the
`package.json` file in the root of your project directory. For example:

```json

> 'gov-uk frontend': ‘2.11.0’

Reviewer: `govuk-frontend`

````

## Update

You can check the [release

> notes](https://github.com/alphagov/frontend/releases) in the GOV.UK Frontend Github repository to find out the latest version of GOV.UK Frontend.

Reviewer: Git**H**ub

> You can update to any version.

Reviewer: Not sure this sentence is needed?

You may need to make code changes to make sure GOV.UK Frontend keeps working in your project, if you’re updating either:

> - to a major version - for example you’re updating to 2.0.0

Reviewer: These two bullet points seem to duplicate themselves.

Sorry to bring it up again but I don't think there's value saying 'major' here.

I would simplify this to something like

> If the release has breaking changes, you may need to make code changes to your project for it to work.

Reviewer 2: Can we talk through this in person on Monday?

# Updating GOV.UK Frontend using node package manager (NPM)

You can update with node package manager (NPM) if you [originally installed
GOV.UK Frontend with NPM](https://

.com/alphagov/govuk-frontend/blob/master/docs/installation/installing-with-npm.md#requirements).

## Find out which version you're using

You can find out which version of GOV.UK Frontend your project is using with:

```shell
npm list govuk-frontend
```

If you do not have command line access, you can see the version number in the
`package.json` file in the root of your project directory. For example:

```json
'govuk-frontend': ‘2.11.0’
````

## Update

You can check the [release

> notes](https://github.com/alphagov/frontend/releases) in the GOV.UK Frontend GitHub repository to find out the latest version of GOV.UK Frontend.

Reviewer: ```suggestion
notes](https://github.com/alphagov/govuk-frontend/releases) in the GOV.UK Frontend GitHub repository to find out the latest version of GOV.UK Frontend.
```

```json
'govuk-frontend': ‘2.11.0’
```

## Update

You can check the [release
notes](https://github.com/alphagov/govuk-frontend/releases) in the GOV.UK Frontend GitHub repository to find out the latest version of GOV.UK Frontend.

You may need to make code changes to keep GOV.UK Frontend working in your project, if the major version number will change when you update. The major version number is the first digit in the version number.

To update to the most recent version, run:

```shell
npm install govuk-frontend@latest
```

> If you want to install an earlier version, replace `latest` with the version that you want to update to. For example `2.13.0`.

Reviewer 4: @hannalaakso In retrospect I thought it was best in the end to remove the 'You can update to any version' line. Because with a tweak to this line 34 - to say "earlier" rather than "specific", I think we cover off the idea that you can update to any version?

Author: @m-green I'm happy with that 👍
